sopel/coretasks.py

```diff
@@ -496,49 +501,88 @@ def track_modes(bot, trigger):
         "Y": module.OPER,
     }
 
-    # Parse modes before doing anything else
-    modes = []
-    sign = ''
+    # Process modes
+    sign = ""
+    param_idx = 0
     for char in modestring:
-        # There was a comment claiming IRC allows e.g. MODE +aB-c foo, but it
-        # doesn't seem to appear in any RFCs. But modern.ircdocs.horse shows
-        # it, so we'll leave in the extra parsing for now.
-        if char in '+-':
+        # Are we setting or unsetting
+        if char in "+-":
             sign = char
-        elif char in mapping:
-            # Filter out unexpected modes and hope they don't have parameters
-            modes.append(sign + char)
-
-    # Try to map modes to arguments, after sanity-checking
-    if len(modes) != len(nicks) or not all([nick.is_nick() for nick in nicks]):
-        # Something fucky happening, like unusual batching of non-privilege
-        # modes together with the ones we expect. Way easier to just re-WHO
-        # than try to account for non-standard parameter-taking modes.
-        LOGGER.debug('Sending WHO for channel: %s', channel)
-        _send_who(bot, channel)
-        return
-
-    for (mode, nick) in zip(modes, nicks):
-        priv = bot.channels[channel].privileges.get(nick, 0)
-        # Log a warning if the two privilege-tracking data structures
-        # get out of sync. That should never happen.
-        # This is a good place to verify that bot.channels is doing
-        # what it's supposed to do before ultimately removing the old,
-        # deprecated bot.privileges structure completely.
-        ppriv = bot.privileges[channel].get(nick, 0)
-        if priv != ppriv:
-            LOGGER.warning("Privilege data error! Please share Sopel's"
-                           "raw log with the developers, if enabled. "
-                           "(Expected {} == {} for {} in {}.)"
-                           .format(priv, ppriv, nick, channel))
-        value = mapping.get(mode[1])
-        if value is not None:
-            if mode[0] == '+':
-                priv = priv | value
-            else:
-                priv = priv & ~value
-            bot.privileges[channel][nick] = priv
-            bot.channels[channel].privileges[nick] = priv
+            continue
+
+        chanmodes = bot.isupport.CHANMODES
+
+        if char in chanmodes["A"]:
```

Comment:
I'd put that before the for-in loop:

```suggestion
            continue

        if char in chanmodes["A"]:
```

Reply:
Well-spotted, thanks